File: scripts/test-migration-guide-contract.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

require 'fileutils'
require 'open3'
require 'tmpdir'

class ContractError < StandardError; end

GUIDE_PATH = 'docs/migration-0.4.md'
EXAMPLES = %w[
  profile-extended
  inline-parser-argument
  options-clone
  fenced-code-pattern
  parse-result-headings
].freeze

def fail_contract(message)
  raise ContractError, "migration guide contract: #{message}"
end

def extract_example(document, name)
  marker = "<!-- migration-example: #{name} -->"
  start = document.index(marker)
  fail_contract("must contain #{marker}") unless start

  block = document[(start + marker.length)..]
  match = block.match(/\A[ \t\r\n]*```rust\r?\n(?<code>.*?)^```[ \t]*\r?$/m)
  fail_contract("#{name} must be followed by a Rust code block") unless match

  match[:code]
end

def assert_contains(document, needle, message)
  fail_contract(message) unless document.include?(needle)
end

def validate_document(document, readme, cargo_toml, node_package, node_workspace, changelog)
  assert_contains(document, '# Migrating from ferromark 0.3 to 0.7', 'must name its supported upgrade range')
  [
    '## Before you start',
    '## 0.4: replace `Profile`',
    '## 0.5: update configurable inline parsing',
    '## 0.6: clone options and make integration matches forward-compatible',
    '## 0.7: raise runtime prerequisites',
    '## Validate the completed upgrade'
  ].each { |heading| assert_contains(document, heading, "must contain #{heading}") }

  assert_contains(readme, "[0.4–0.7 migration guide](#{GUIDE_PATH})", 'README must link the 0.4–0.7 guide')
  assert_contains(document, 'Options::from(Profile)', 'must explain the removed Profile conversion')
  assert_contains(document, 'inline_footnotes: bool', 'must name the inserted inline-parser argument')
  assert_contains(document, 'InlineFootnote(Range)', 'must name the added inline event')
  assert_contains(document, 'Options` no longer implements `Copy`', 'must explain the Options Copy removal')
  assert_contains(document, '`#[non_exhaustive]`', 'must explain the fenced-code forward-compatibility contract')
  assert_contains(document, '`headings`', 'must explain the ParseResult metadata field')
  [
    'remove Profile and Options::from(Profile)',
    'a positional parse_with_options argument',
    'Options no longer implements Copy',
    'FencedCodeBlock gained the meta field and is now non_exhaustive',
    'ferromark now requires Rust 1.88 or newer',
    'ferromark npm package now requires Node.js 22 or newer'
  ].each { |change| assert_contains(changelog, change, "CHANGELOG must record #{change.inspect}") }

  rust_version = cargo_toml[/^rust-version\s*=\s*"([^"]+)"\s*$/, 1]
  fail_contract('Cargo.toml must declare rust-version') unless rust_version
  assert_contains(document, "requires Rust #{rust_version} or newer", 'must state Cargo.toml Rust requirement')

  node_version = node_package[/"node"\s*:\s*">=([0-9]+(?:\.[0-9]+){0,2})"/, 1]
  fail_contract('node package must declare a minimum Node version') unless node_version
  assert_contains(document, "Node.js #{node_version} or newer", 'must state npm package Node requirement')

  workspace_node = node_workspace[/"node"\s*:\s*">=([0-9]+(?:\.[0-9]+){0,2})"/, 1]
  workspace_pnpm = node_workspace[/"packageManager"\s*:\s*"pnpm@([^"]+)"/, 1]
  fail_contract('node workspace must declare a minimum Node version') unless workspace_node
  fail_contract('node workspace must pin pnpm') unless workspace_pnpm
  assert_contains(document, "pnpm #{workspace_pnpm}", 'must state the node workspace pnpm version')
  assert_contains(document, "Node.js #{workspace_node} or newer", 'must state the node workspace Node requirement')

  EXAMPLES.each { |name| extract_example(document, name) }
  assert_contains(
    extract_example(document, 'inline-parser-argument'),
    'false, // inline_footnotes',
    'inline parser example must preserve the new argument position'
  )
  assert_contains(
    extract_example(document, 'fenced-code-pattern'),
    '..',
    'fenced-code example must ignore future fields'
  )
end

def compile_examples(repository_root, document)
  examples = EXAMPLES.map { |name| [name, extract_example(document, name)] }

  Dir.mktmpdir('ferromark-migration-guide.') do |directory|
    File.write(File.join(directory, 'Cargo.toml'), <<~TOML)
      [package]
      name = "ferromark-migration-guide-contract"
      version = "0.0.0"
      edition = "2024"

      [dependencies]
      ferromark = { path = #{repository_root.inspect}, features = ["mdx"] }
    TOML
    FileUtils.mkdir_p(File.join(directory, 'src'))
    functions = examples.map do |name, code|
      "fn #{name.tr('-', '_')}() {\n#{code.lines.map { |line| "  #{line}" }.join}\n}\n"
    end.join("\n")
    File.write(File.join(directory, 'src/main.rs'), "#![allow(dead_code)]\n\n#{functions}\nfn main() {}\n")

    stdout, stderr, status = Open3.capture3(
      'cargo', 'check', '--quiet', '--manifest-path', File.join(directory, 'Cargo.toml')
    )
    next if status.success?

    fail_contract("documented Rust examples must compile:\n#{stdout}#{stderr}")
  end
end

def validate(repository_root)
  guide = File.read(File.join(repository_root, GUIDE_PATH))
  readme = File.read(File.join(repository_root, 'README.md'))
  cargo_toml = File.read(File.join(repository_root, 'Cargo.toml'))
  node_package = File.read(File.join(repository_root, 'node/ferromark/package.json'))
  node_workspace = File.read(File.join(repository_root, 'node/package.json'))
  changelog = File.read(File.join(repository_root, 'CHANGELOG.md'))
  validate_document(guide, readme, cargo_toml, node_package, node_workspace, changelog)
  compile_examples(repository_root, guide)
end

def assert_rejected(label)
  yield
  raise "#{label} mutation unexpectedly passed"
rescue ContractError
  # Expected: this mutation must be rejected by the contract.
end

def self_test(repository_root)
  validate(repository_root)

  guide = File.read(File.join(repository_root, GUIDE_PATH))
  readme = File.read(File.join(repository_root, 'README.md'))
  cargo_toml = File.read(File.join(repository_root, 'Cargo.toml'))
  node_package = File.read(File.join(repository_root, 'node/ferromark/package.json'))
  node_workspace = File.read(File.join(repository_root, 'node/package.json'))
  changelog = File.read(File.join(repository_root, 'CHANGELOG.md'))

  assert_rejected('README guide link') do
    validate_document(
      guide,
      readme.sub("[0.4–0.7 migration guide](#{GUIDE_PATH})", 'migration guide'),
      cargo_toml,
      node_package,
      node_workspace,
      changelog
    )
  end
  assert_rejected('inline parser argument') do
    validate_document(guide.sub('false, // inline_footnotes', 'None, // footnote_store'), readme, cargo_toml, node_package, node_workspace, changelog)
  end
  assert_rejected('fenced-code fallback pattern') do
    validate_document(guide.sub("            ..\n", ''), readme, cargo_toml, node_package, node_workspace, changelog)
  end
  assert_rejected('Rust version') do
    validate_document(guide.sub('requires Rust 1.88 or newer', 'requires Rust 1.85 or newer'), readme, cargo_toml, node_package, node_workspace, changelog)
  end
  assert_rejected('Node version') do
    validate_document(guide.sub('Node.js 22.12.0 or newer', 'Node.js 20.0.0 or newer'), readme, cargo_toml, node_package, node_workspace, changelog)
  end
end

repository_root = File.expand_path('..', __dir__)

begin
  if ARGV == ['--self-test']
    self_test(repository_root)
  elsif ARGV.empty?
    validate(repository_root)
  else
    abort 'usage: test-migration-guide-contract.rb [--self-test]'
  end
rescue ContractError => error
  abort error.message
end

puts 'Migration guide contract checks passed'

File: scripts/test-migration-guide-contract.mjs
import assert from 'node:assert/strict'
import { spawnSync } from 'node:child_process'
import { mkdirSync, mkdtempSync, rmSync, writeFileSync } from 'node:fs'
import { tmpdir } from 'node:os'
import path from 'node:path'
import { describe, it } from 'node:test'

import { ContractError, readRepositoryFile, repositoryRoot } from './lib/contracts.mjs'

const GUIDE_PATH = 'docs/migration-0.4.md'
const EXAMPLES = [
  'profile-extended',
  'inline-parser-argument',
  'options-clone',
  'fenced-code-pattern',
  'parse-result-headings',
]

function failContract(message) {
  throw new ContractError(`migration guide contract: ${message}`)
}

function extractExample(document, name) {
  const marker = `<!-- migration-example: ${name} -->`
  const start = document.indexOf(marker)
  if (start === -1) {
    failContract(`must contain ${marker}`)
  }
  const block = document.slice(start + marker.length)
  const match = block.match(/^[ \t\r\n]*```rust\r?\n([\s\S]*?)^```[ \t]*\r?$/m)
  if (!match) {
    failContract(`${name} must be followed by a Rust code block`)
  }
  return match[1]
}

function assertContains(document, needle, message) {
  if (!document.includes(needle)) {
    failContract(message)
  }
}

function validateDocument({ guide, readme, cargoToml, nodePackage, nodeWorkspace, changelog }) {
  assertContains(guide, '# Migrating from ferromark 0.3 to 0.7', 'must name its supported upgrade range')
  for (const heading of [
    '## Before you start',
    '## 0.4: replace `Profile`',
    '## 0.5: update configurable inline parsing',
    '## 0.6: clone options and make integration matches forward-compatible',
    '## 0.7: raise runtime prerequisites',
    '## Validate the completed upgrade',
  ]) {
    assertContains(guide, heading, `must contain ${heading}`)
  }

  assertContains(
    readme,
    `[0.4–0.7 migration guide](${GUIDE_PATH})`,
    'README must link the 0.4–0.7 guide',
  )
  assertContains(guide, 'Options::from(Profile)', 'must explain the removed Profile conversion')
  assertContains(guide, 'inline_footnotes: bool', 'must name the inserted inline-parser argument')
  assertContains(guide, 'InlineFootnote(Range)', 'must name the added inline event')
  assertContains(guide, 'Options` no longer implements `Copy`', 'must explain the Options Copy removal')
  assertContains(guide, '`#[non_exhaustive]`', 'must explain the fenced-code forward-compatibility contract')
  assertContains(guide, '`headings`', 'must explain the ParseResult metadata field')
  for (const change of [
    'remove Profile and Options::from(Profile)',
    'a positional parse_with_options argument',
    'Options no longer implements Copy',
    'FencedCodeBlock gained the meta field and is now non_exhaustive',
    'ferromark now requires Rust 1.88 or newer',
    'ferromark npm package now requires Node.js 22 or newer',
  ]) {
    assertContains(changelog, change, `CHANGELOG must record ${JSON.stringify(change)}`)
  }

  const rustVersion = cargoToml.match(/^rust-version\s*=\s*"([^"]+)"\s*$/m)?.[1]
  if (!rustVersion) {
    failContract('Cargo.toml must declare rust-version')
  }
  assertContains(
    guide,
    `requires Rust ${rustVersion} or newer`,
    'must state Cargo.toml Rust requirement',
  )

  const nodeVersion = nodePackage.match(/"node"\s*:\s*">=([0-9]+(?:\.[0-9]+){0,2})"/)?.[1]
  if (!nodeVersion) {
    failContract('node package must declare a minimum Node version')
  }
  assertContains(guide, `Node.js ${nodeVersion} or newer`, 'must state npm package Node requirement')

  const workspaceNode = nodeWorkspace.match(/"node"\s*:\s*">=([0-9]+(?:\.[0-9]+){0,2})"/)?.[1]
  const workspacePnpm = nodeWorkspace.match(/"packageManager"\s*:\s*"pnpm@([^"]+)"/)?.[1]
  if (!workspaceNode) {
    failContract('node workspace must declare a minimum Node version')
  }
  if (!workspacePnpm) {
    failContract('node workspace must pin pnpm')
  }
  assertContains(guide, `pnpm ${workspacePnpm}`, 'must state the node workspace pnpm version')
  assertContains(
    guide,
    `Node.js ${workspaceNode} or newer`,
    'must state the node workspace Node requirement',
  )

  for (const name of EXAMPLES) {
    extractExample(guide, name)
  }
  assertContains(
    extractExample(guide, 'inline-parser-argument'),
    'false, // inline_footnotes',
    'inline parser example must preserve the new argument position',
  )
  assertContains(
    extractExample(guide, 'fenced-code-pattern'),
    '..',
    'fenced-code example must ignore future fields',
  )
}

function compileExamples(guide) {
  const directory = mkdtempSync(path.join(tmpdir(), 'ferromark-migration-guide.'))
  try {
    writeFileSync(
      path.join(directory, 'Cargo.toml'),
      `[package]\nname = "ferromark-migration-guide-contract"\nversion = "0.0.0"\nedition = "2024"\n\n[dependencies]\nferromark = { path = ${JSON.stringify(repositoryRoot)}, features = ["mdx"] }\n`,
    )
    mkdirSync(path.join(directory, 'src'), { recursive: true })
    const functions = EXAMPLES.map(name => {
      const code = extractExample(guide, name)
        .split('\n')
        .map(line => `  ${line}`)
        .join('\n')
      return `fn ${name.replaceAll('-', '_')}() {\n${code}\n}\n`
    }).join('\n')
    writeFileSync(
      path.join(directory, 'src/main.rs'),
      `#![allow(dead_code)]\n\n${functions}\nfn main() {}\n`,
    )

    const result = spawnSync(
      'cargo',
      ['check', '--quiet', '--manifest-path', path.join(directory, 'Cargo.toml')],
      { encoding: 'utf8' },
    )
    if (result.error) {
      failContract(`cargo check could not be started: ${result.error.message}`)
    }
    if (result.status !== 0) {
      failContract(`documented Rust examples must compile:\n${result.stdout}${result.stderr}`)
    }
  } finally {
    rmSync(directory, { force: true, recursive: true })
  }
}

const inputs = {
  guide: readRepositoryFile(GUIDE_PATH),
  readme: readRepositoryFile('README.md'),
  cargoToml: readRepositoryFile('Cargo.toml'),
  nodePackage: readRepositoryFile('node/ferromark/package.json'),
  nodeWorkspace: readRepositoryFile('node/package.json'),
  changelog: readRepositoryFile('CHANGELOG.md'),
}

function assertRejected(overrides) {
  assert.throws(() => validateDocument({ ...inputs, ...overrides }), ContractError)
}

describe('migration guide contract', () => {
  it('accepts the current guide', () => {
    validateDocument(inputs)
  })

  it('compiles every documented example', () => {
    compileExamples(inputs.guide)
  })

  it('rejects a missing README guide link', () => {
    assertRejected({
      readme: inputs.readme.replace(`[0.4–0.7 migration guide](${GUIDE_PATH})`, 'migration guide'),
    })
  })

  it('rejects a stale inline parser argument', () => {
    assertRejected({
      guide: inputs.guide.replace('false, // inline_footnotes', 'None, // footnote_store'),
    })
  })

  it('rejects a fenced-code example without the fallback pattern', () => {
    assertRejected({ guide: inputs.guide.replace('            ..\n', '') })
  })

  it('rejects a stale Rust version', () => {
    assertRejected({
      guide: inputs.guide.replace('requires Rust 1.94 or newer', 'requires Rust 1.85 or newer'),
    })
  })

  it('rejects a stale Node version', () => {
    assertRejected({
      guide: inputs.guide.replace('Node.js 22.12.0 or newer', 'Node.js 20.0.0 or newer'),
    })
  })
})
